In [1]:
import sys
import os
sys.path.append("./AIMA/")
from knights2 import *

In [3]:
#Example of a game step by step
play_game(Knights_Move(), alphabeta_player(1), alphabeta_player(1))

INITIAL BOARD
\ 1 2 3 4 5 6 7 8 /
1 . . . . . . . . 1
2 . . . . . . . . 2
3 . . . . . . 2 . 3
4 . . . . . . . . 4
5 . . . . . . . . 5
6 . . . . . 1 . . 6
7 . . . . . . . . 7
8 . . . . . . . . 8
/ 1 2 3 4 5 6 7 8 \
PLAY # 1
legal moves= [(8, 5), (8, 7), (7, 8), (5, 8), (4, 7), (4, 5), (5, 4), (7, 4)]
Player 1 selects (8, 5)
\ 1 2 3 4 5 6 7 8 /
1 . . . . . . . . 1
2 . . . . . . . . 2
3 . . . . . . 2 . 3
4 . . . . . . . . 4
5 . . . . . . . . 5
6 . . . . . _ . . 6
7 . . . . . . . . 7
8 . . . . 1 . . . 8
/ 1 2 3 4 5 6 7 8 \
legal moves= [(5, 6), (5, 8), (1, 8), (1, 6), (2, 5), (4, 5)]
Player 2 selects (5, 6)
\ 1 2 3 4 5 6 7 8 /
1 . . . . . . . . 1
2 . . . . . . . . 2
3 . . . . . . _ . 3
4 . . . . . . . . 4
5 . . . . . 2 . . 5
6 . . . . . _ . . 6
7 . . . . . . . . 7
8 . . . . 1 . . . 8
/ 1 2 3 4 5 6 7 8 \
PLAY # 2
legal moves= [(7, 7), (6, 4), (7, 3)]
Player 1 selects (7, 7)
\ 1 2 3 4 5 6 7 8 /
1 . . . . . . . . 1
2 . . . . . . . . 2
3 . . . . . . _ . 3
4 . . . . . . . . 4
5 . . . . . 2 . . 

5 . _ . _ _ _ _ _ 5
6 . . . _ _ _ _ _ 6
7 _ . _ 1 _ _ _ _ 7
8 . . _ _ _ _ _ _ 8
/ 1 2 3 4 5 6 7 8 \
legal moves= [(1, 8), (1, 4), (3, 4)]
Player 2 selects (1, 8)
\ 1 2 3 4 5 6 7 8 /
1 . . . . . _ _ 2 1
2 . . . . . _ . _ 2
3 . . . . _ _ _ _ 3
4 . . . . _ _ _ . 4
5 . _ . _ _ _ _ _ 5
6 . . . _ _ _ _ _ 6
7 _ . _ 1 _ _ _ _ 7
8 . . _ _ _ _ _ _ 8
/ 1 2 3 4 5 6 7 8 \
PLAY # 18
legal moves= [(5, 3), (6, 2), (8, 2)]
Player 1 selects (5, 3)
END BOARD
\ 1 2 3 4 5 6 7 8 /
1 . . . . . _ _ 2 1
2 . . . . . _ . _ 2
3 . . . . _ _ _ _ 3
4 . . . . _ _ _ . 4
5 . _ 1 _ _ _ _ _ 5
6 . . . _ _ _ _ _ 6
7 _ . _ _ _ _ _ _ 7
8 . . _ _ _ _ _ _ 8
/ 1 2 3 4 5 6 7 8 \
Player 1 wins!!


1

In [59]:
#Example of Conant's state (this one is an initial state)
N = 8
moves = [(x, y) for x in range(1, N+1) for y in range(1, N+1)]
knight1 = random.choice(moves)
moves.remove(knight1)
knight2 = random.choice(moves)
moves.remove(knight2)
#variable created containing who moves next, the location of the knights and ALL cells of the board
initialization = GameState(to_move=1, knights={1:knight1, 2:knight2}, moves=set(moves))
print(initialization)

#Aterwards ALL cells of the board are filtered according to who moves next and how the knight can move
moves = Knights_Move().actions(initialization) #Returns legal moves of knight1 in this example
print("\n" + str(moves))
Knights_Move().result(initialization, moves[0]) #Applies an action to the state and generates hypothetical new state

GameState(to_move=1, knights={1: (6, 8), 2: (3, 1)}, moves={(7, 3), (4, 7), (1, 3), (4, 8), (5, 6), (2, 8), (6, 6), (7, 7), (2, 1), (6, 2), (1, 6), (3, 7), (2, 5), (5, 1), (8, 5), (5, 8), (7, 2), (1, 2), (6, 7), (3, 3), (5, 5), (8, 1), (7, 6), (4, 4), (6, 3), (1, 5), (3, 6), (2, 2), (8, 6), (4, 1), (1, 1), (6, 4), (3, 2), (2, 6), (5, 4), (8, 2), (7, 1), (4, 5), (5, 2), (1, 4), (7, 5), (2, 3), (8, 7), (4, 2), (6, 5), (3, 5), (2, 7), (5, 3), (7, 8), (8, 3), (4, 6), (6, 1), (5, 7), (3, 8), (7, 4), (1, 8), (8, 8), (4, 3), (1, 7), (3, 4), (2, 4), (8, 4)})

[(8, 7), (4, 7), (5, 6), (7, 6)]


GameState(to_move=2, knights={1: (8, 7), 2: (3, 1)}, moves={(7, 3), (4, 7), (1, 3), (4, 8), (5, 6), (2, 8), (6, 6), (7, 7), (2, 1), (6, 2), (1, 6), (3, 7), (2, 5), (5, 1), (8, 5), (5, 8), (7, 2), (1, 2), (6, 7), (3, 3), (5, 5), (8, 1), (7, 6), (4, 4), (6, 3), (1, 5), (3, 6), (2, 2), (8, 6), (4, 1), (1, 1), (6, 4), (3, 2), (2, 6), (5, 4), (8, 2), (7, 1), (4, 5), (5, 2), (1, 4), (7, 5), (2, 3), (4, 2), (6, 5), (3, 5), (2, 7), (5, 3), (7, 8), (8, 3), (4, 6), (6, 1), (5, 7), (3, 8), (7, 4), (1, 8), (8, 8), (4, 3), (1, 7), (3, 4), (2, 4), (8, 4)})

In [2]:
#Just an utility to not print stuff in the loop experiments
class HiddenPrints:
    def __enter__(self):
        self._original_stdout = sys.stdout
        sys.stdout = open(os.devnull, 'w')

    def __exit__(self, exc_type, exc_val, exc_tb):
        sys.stdout.close()
        sys.stdout = self._original_stdout

In [19]:
#Testing how many games out of 10000 we win without changing the default evaluation function
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(1), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with default evaluation function: " + str(wins))
print("(Lost games: " + str(loses) + ")")

Won games with default evaluation function: 4993
(Lost games: 5006)


We could say this is like a 50/50 chance of winning more or less.
It depends quite a bit on the random starting positions

In [8]:
#Defining the evaluation function and 1 feature
def smartKnight(state, player):
    w_1 = 1
    f_1 = defend(state, player)
    
    evaluation = w_1 * f_1
    return evaluation

def defend(state, player):
    '''Heuristic that evaluates the position of both max a min according to a cell-point system (move_constraints.png).
    Returns the difference of the max defense value and min defense value'''
    
    if player == 1: #Max is knight1
        player_position = prs = state[1][1]
        opponent_position = ops = state[1][2]
    else: #Max is knight2
        player_position = prs = state[1][2]
        opponent_position = ops = state[1][1]
    
    #MAX DEFENSE CALCULATION
    if prs[0] == 1 or prs[0] == 8: #Max on top or bottom edge of the board
        if prs[1] == 1 or prs[1] == 8: #Max in a corner of the board
            prs_def = 2
        elif prs[1] == 2 or prs[1] == 7: #Max close to the corner of the board
            prs_def = 3
        else: #Max on edge of the board but not close to the corners
            prs_def = 4
    
    elif prs[0] == 2 or prs[0] == 7:
        if prs[1] == 1 or prs[1] == 8: #Max close to the corner of the board
            prs_def = 3
        elif ops[1] == 2 or ops[1] == 7: #Max on edge of the board but not close to the corners
            prs_def = 4
        else: #Max on the peripheria of the board center
            prs_def = 5
    
    else:
        if prs[1] == 1 or prs[1] == 8: #Max on edge of the board but not close to the corners
            prs_def = 4
        elif prs[1] == 2 or prs[1] == 7: #Max on the peripheria of the board center
            prs_def = 5
        else: #Max in the central 4x4 area of the board
            prs_def = 8
    
    #MIN DEFENSE CALCULATION (same logic as max)
    if ops[0] == 1 or ops[0] == 8:
        if ops[1] == 1 or ops[1] == 8:
            ops_def = 2
        elif ops[1] == 2 or ops[1] == 7:
            ops_def = 3
        else:
            ops_def = 4
    
    elif ops[0] == 2 or ops[0] == 7:
        if ops[1] == 1 or ops[1] == 8:
            ops_def = 3
        elif ops[1] == 2 or ops[1] == 7:
            ops_def = 4
        else:
            ops_def = 5
    
    else:
        if ops[1] == 1 or ops[1] == 8:
            ops_def = 4
        elif ops[1] == 2 or ops[1] == 7:
            ops_def = 5
        else:
            ops_def = 8
    
    return prs_def - ops_def

In [4]:
#Testing how many games out of 10000 we win with our evaluation function with defense as its only feature
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1: 
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (only defense heuristic): " + str(wins))
print("(Lost games: " + str(loses) + ")")

Won games with smartKnight function (only escape heuristic): 7450
(Lost games: 2549)


This seemingly increments our chances of winning!
In fact when we compute a one tail proportions z-test in R:

`prop.test( x = c(7450, 4993), n = c(10000, 10000), alternative = "greater" )$p.value`

We get a very small p of `2.856481e-281`, in general we can establish that a p value `p <= 0.05` rejects the null hypothesis that both binomial distributions are equal. This indicates that the probability that we got significantly more wins with the defend heuristic than with the default evaluation function is really high.

In [12]:
#Testing another standalone heuristic
def smartKnight(state, player):
    w_2 = 1
    f_2 = attack(state, player)
    
    evaluation = w_2 * f_2
    return evaluation

def attack(state, player):
    '''Heuristic that evaluates how offensive a knight is. Returns the difference of the max attack
    valu and the min attack value.'''
    
    if player == 1: #Max is knight1
        player_position = prs = state[1][1]
        opponent_position = ops = state[1][2]
    else: #Max is knight2
        player_position = prs = state[1][2]
        opponent_position = ops = state[1][1]
    
    #Check if agression exists, agressions are always bilateral because of how knights move
    agression = False
    if prs[0] == ops[0] + 1 or prs[0] == ops[0] - 1:
        if prs[1] == ops[1] + 2 or prs[1] == ops[1] - 2:
            agression = True #Knights attacking each other in "short L" pattern
    elif prs[0] == ops[0] + 2 or prs[0] == ops[0] - 2:
        if prs[1] == ops[1] + 1 or prs[1] == ops[1] - 1:
            agression = True #Knights attacking each other in "tall L" pattern
    
    #COMPUTE MAX'S ATTACK VALUE
    if agression == True:
        if ops[0] == 1 or ops[0] == 8: #Min is vulnerable on the edge of the board
            prs_attack = 2
            if ops[1] == 1 or ops[1] == 8: #Min is trapped in the corner of the board
                prs_attack = 3
        elif ops[1] == 1 or ops[1] == 8: #Min is vulnerable on the edge of the board
            prs_attack = 2
        else: #The has a good chance of being unimportant as it is not happening in the edge of the board
            prs_attack = 0
    
    #COMPUTE MIN'S ATTACK VALUE (same logic as max)
    if agression == True:
        if prs[0] == 1 or prs[0] == 8:
            ops_attack = 2
            if prs[1] == 1 or prs[1] == 8:
                ops_attack = 3
        elif prs[1] == 1 or prs[1] == 8:
            ops_attack = 2
        else:
            ops_attack = 0
    
    if agression == True:
        return prs_attack - ops_attack
    else: #Indirectly promote agression if it doesn't exits yet
        return -1

In [6]:
#Testing how many games out of 10000 we win with our evaluation function with one feature (attack heuristic)
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (attack heuristic): " + str(wins))
print("(Lost games: " + str(loses) + ")")

Won games with smartKnight function (attack heuristic): 6205
(Lost games: 3794)


There seems to be an improvement of wins here over just using the default evaluation function. When the two proportions are tested in R:

`prop.test( x = c(6205, 4993), n = c(10000, 10000), alternative = "greater" )$p.value`

using the one-tailed z-test we get a p value of `5.559969e-67` which indicates that the attack heuristic does help win more games. Now lets try combining the two heuristics we have so far:

In [10]:
#Defining the evaluation function with 2 features
def smartKnight(state, player):
    w_1 = 1
    f_1 = defend(state, player)
    
    w_2 = 1
    f_2 = attack(state, player)
    
    evaluation = w_1*f_1 + w_2*f_2
    return evaluation

In [11]:
#Testing how many games out of 10000 we win with our evaluation function with the two features combined
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (combined heuristics): " + str(wins))
print("(Lost games: " + str(loses) + ")")

Won games with smartKnight function (combined heuristics): 8058
(Lost games: 1941)


This time we will perform the statistical test against the performance of the defend heuristic alone which was our best result so far:

`prop.test( x = c(8058, 7450), n = c(10000, 10000), alternative = "greater" )$p.value`

This yields a p value of `4.115462e-25` which strongly indicates that this combination of heuristics is better than just defend alone. However we think it might be possible to improve the performance of this evaluation function by tuning the weight of the attack heuristic under the intuition presented in "attack.png":

In [17]:
#Defining the evaluation function with 2 features and tuned weight for heuristic attack
def smartKnight(state, player):
    w_1 = 1
    f_1 = defend(state, player)
    
    w_2 = 1.5
    f_2 = attack(state, player)
    
    evaluation = w_1*f_1 + w_2*f_2
    return evaluation

In [18]:
#Testing how many games out of 10000 we win with our evaluation function with the two features combined and tuning
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (combined heuristics with tuned attack): " + str(wins))
print("(Lost games: " + str(loses) + ")")

Won games with smartKnight function (combined heuristics with tuned attack): 8105
(Lost games: 1894)


Let us test this apparent improvent against the last experiment without tuning the parameter:

`prop.test( x = c(8058, 8105), n = c(10000, 10000), alternative = "greater" )$p.value`

The result is a p-value of `0.7956173` which statistically indicates that the apparent performance improvement is not significant. However since the tuned weight doesn't negatively affect the agent and it is concordant with our intuition we are considering leaving it in with the adjustement.

In [49]:
def smartKnight(state, player):
    w_3 = 1
    f_3 = stall(state, player)
    
    evaluation = w_3*f_3
    return evaluation

def stall(state, player):
    if player == 1: #Max is knight1
        player_position = prs = state[1][1]
        opponent_position = ops = state[1][2]
    else: #Max is knight2
        player_position = prs = state[1][2]
        opponent_position = ops = state[1][1]
    
    prs_center = (prs[0] - 4)^2 + (prs[1] - 4)^2
    ops_center = (ops[0] - 4)^2 + (ops[1] - 4)^2
    
    return prs_center - ops_center

In [50]:
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (stall heuristic): " + str(wins))
print("(Lost games: " + str(loses) + ")")

Won games with smartKnight function (stall heuristic): 7762
(Lost games: 2237)


prop.test( x = c(5391, 4993), n = c(10000, 10000), alternative = "greater" )$p.value

In [45]:
#Defining the evaluation function and 1 feature
def smartKnight(state, player):
    w_3 = 1
    f_3 = defendSquare(state, player)
    
    evaluation = w_3 * f_3
    return evaluation

def defendSquare(state, player):
    '''Heuristic that evaluates the position of both max a min according to a cell-point system (move_constraints.png).
    Returns the difference of the max defense value and min defense value'''
    
    if player == 1: #Max is knight1
        player_position = prs = state[1][1]
        opponent_position = ops = state[1][2]
    else: #Max is knight2
        player_position = prs = state[1][2]
        opponent_position = ops = state[1][1]
    
    #MAX DEFENSE CALCULATION
    if prs[0] == 1 or prs[0] == 8: #Max on top or bottom edge of the board
        if prs[1] == 1 or prs[1] == 8: #Max in a corner of the board
            prs_def = 2
        elif prs[1] == 2 or prs[1] == 7: #Max close to the corner of the board
            prs_def = 3
        else: #Max on edge of the board but not close to the corners
            prs_def = 4
    
    elif prs[0] == 2 or prs[0] == 7:
        if prs[1] == 1 or prs[1] == 8: #Max close to the corner of the board
            prs_def = 3
        elif ops[1] == 2 or ops[1] == 7: #Max on edge of the board but not close to the corners
            prs_def = 4
        else: #Max on the peripheria of the board center
            prs_def = 5
    
    else:
        if prs[1] == 1 or prs[1] == 8: #Max on edge of the board but not close to the corners
            prs_def = 4
        elif prs[1] == 2 or prs[1] == 7: #Max on the peripheria of the board center
            prs_def = 5
        else: #Max in the central 4x4 area of the board
            prs_def = 8
    
    #MIN DEFENSE CALCULATION (same logic as max)
    if ops[0] == 1 or ops[0] == 8:
        if ops[1] == 1 or ops[1] == 8:
            ops_def = 2
        elif ops[1] == 2 or ops[1] == 7:
            ops_def = 3
        else:
            ops_def = 4
    
    elif ops[0] == 2 or ops[0] == 7:
        if ops[1] == 1 or ops[1] == 8:
            ops_def = 3
        elif ops[1] == 2 or ops[1] == 7:
            ops_def = 4
        else:
            ops_def = 5
    
    else:
        if ops[1] == 1 or ops[1] == 8:
            ops_def = 4
        elif ops[1] == 2 or ops[1] == 7:
            ops_def = 5
        else:
            ops_def = 8
    
    return prs_def - (2 * ops_def)

In [46]:
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (stall heuristic): " + str(wins))
print("(Lost games: " + str(loses) + ")")

Won games with smartKnight function (stall heuristic): 7464
(Lost games: 2535)


In [55]:
#Defining the evaluation function with 2 features and tuned weight for heuristic attack
def smartKnight(state, player):
    w_1 = 1
    f_1 = defend(state, player)
    
    w_2 = 1.5
    f_2 = attack(state, player)
    
    w_3 = 0.5 #8432 for 1
    f_3 = stall(state, player)
    
    evaluation = w_1*f_1 + w_2*f_2 + w_3*f_3
    return evaluation

In [ ]:
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (stall heuristic): " + str(wins))
print("(Lost games: " + str(loses) + ")")

`prop.test( x = c(8432, 8105), n = c(10000, 10000), alternative = "greater" )$p.value`

In [ ]:
#Defining the evaluation function with 2 features and tuned weight for heuristic attack
def smartKnight(state, player):
    w_1 = 0.745
    f_1 = defend(state, player)
    
    w_2 = 0.6205
    f_2 = attack(state, player)
    
    w_3 = 0.7464
    f_3 = stall(state, player)
    
    evaluation = w_1*f_1 + w_2*f_2 + w_3*f_3
    return evaluation

In [ ]:
wins = 0
loses = 0
with HiddenPrints():
    for i in range(1, 10000):
        result = play_game(Knights_Move(), alphabeta_player(depth=1, eval_fn=smartKnight), alphabeta_player(1))
        if result == 1:
            wins += 1
        else:
            loses += 1
print("Won games with smartKnight function (stall heuristic): " + str(wins))
print("(Lost games: " + str(loses) + ")")